lib/topic_query.rb

```diff
@@ -897,24 +897,42 @@ def remove_muted_categories(list, user, opts = nil)
     category_id = get_category_id(opts[:exclude]) if opts
 
     if user
+      watched_tag_ids =
+        if SiteSetting.watched_precedence_over_muted
+          TagUser
+            .where(user: user)
+            .where("notification_level >= ?", TopicUser.notification_levels[:watching])
+            .pluck(:tag_id)
+        else
+          []
+        end
+
+      # OR watched_topic_tags.id IS NOT NULL",
       list =
-        list
-          .references("cu")
-          .joins(
-            "LEFT JOIN category_users ON category_users.category_id = topics.category_id AND category_users.user_id = #{user.id}",
+        list.references("cu").joins(
+          "LEFT JOIN category_users ON category_users.category_id = topics.category_id AND category_users.user_id = #{user.id}",
+        )
+      if watched_tag_ids.present?
+        list =
+          list.joins(
+            "LEFT JOIN topic_tags watched_topic_tags ON topic_tags.topic_id = topics.id AND #{DB.sql_fragment("topic_tags.tag_id IN (?)", watched_tag_ids)}",
```

Comment:
```suggestion
          list.joins(
            "LEFT JOIN topic_tags watched_topic_tags ON topic_tags.topic_id = topics.id"
          ).where("topic_tags.tag_id IN (?)", watched_tag_ids)
```

I feel like using `DB.sql_fragment` here makes the query much harder to read. Instead, you can just add a `where` clause which will generate the same query.